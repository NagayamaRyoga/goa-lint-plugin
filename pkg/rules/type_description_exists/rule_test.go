package type_description_exists_test

import (
	"log"
	"os"
	"testing"

	"github.com/NagayamaRyoga/goalint/pkg/rules/type_description_exists"
	"github.com/gkampitakis/go-snaps/snaps"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
	"goa.design/goa/v3/dsl"
	"goa.design/goa/v3/eval"
	"goa.design/goa/v3/expr"
)

func TestRule(t *testing.T) {
	t.Parallel()

	var (
		typeWithDescription = dsl.Type("User", func() {
			dsl.Description("User information")
			dsl.Attribute("id", dsl.Int)
			dsl.Attribute("name", dsl.String)
		})

		resultTypeWithDescription = dsl.ResultType("application/vnd.user", func() {
			dsl.Description("User information")
			dsl.Attribute("id", dsl.Int)
			dsl.Attribute("name", dsl.String)
		})

		typeWithoutDescription = dsl.Type("Book", func() {
			dsl.Attribute("id", dsl.Int)
			dsl.Attribute("title", dsl.String)
		})

		resultTypeWithoutDescription = dsl.ResultType("application/vnd.book", func() {
			dsl.Attribute("id", dsl.Int)
			dsl.Attribute("title", dsl.String)
		})
	)

	// given
	err := eval.RunDSL()
	require.NoError(t, err)

	testCases := []struct {
		description string
		dataType    expr.DataType
		wantReports int
	}{
		{
			description: "success/Type",
			dataType:    typeWithDescription,
			wantReports: 0,
		},
		{
			description: "success/ResultType",
			dataType:    resultTypeWithDescription,
			wantReports: 0,
		},
		{
			description: "failed/Type",
			dataType:    typeWithoutDescription,
			wantReports: 1,
		},
		{
			description: "failed/ResultType",
			dataType:    resultTypeWithoutDescription,
			wantReports: 1,
		},
	}
	for _, tc := range testCases {
		t.Run(tc.description, func(t *testing.T) {
			t.Parallel()

			snaps := snaps.WithConfig(snaps.Dir("testdata"))

			logger := log.New(os.Stdout, "", 0)
			cfg := type_description_exists.NewConfig()
			rule := type_description_exists.NewRule(logger, cfg)

			// when
			got := rule.WalkType(tc.dataType)
			assert.Len(t, got, tc.wantReports)
			snaps.MatchSnapshot(t, got.String())
		})
	}
}
