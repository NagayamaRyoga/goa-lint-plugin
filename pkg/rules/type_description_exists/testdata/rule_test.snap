
[TestRule/success/Type - 1]

---

[TestRule/failed/Type - 1]
Found 1 errors and 0 warnings
[TypeDescriptionExists]: error in Type("Book"): Type should have non-empty description

---

[TestRule/success/ResultType - 1]

---

[TestRule/failed/ResultType - 1]
Found 1 errors and 0 warnings
[TypeDescriptionExists]: error in ResultType("application/vnd.book"): ResultType should have non-empty description

---
